File: Chatbots/Python/Source/Scoring/keyword.py
from __future__ import annotations

from typing import Dict, List, Set, Tuple

from ..dataModel import (
    Card,
    PreparedQuestion,
    AnswerHit,
    deck_path_to_string,
    short_preview,
)
from ..config import ParserConfig
from ..normalise import normalise_for_matching
from ..tokenise import tokenise_to_set


def build_guid_index(cards: List[Card]) -> Dict[str, Card]:
    """Create a guid -> Card lookup for fast result materialisation."""
    return {card.guid: card for card in cards}


def prepare_keyword_index(
    candidate_cards: List[Card],
    stopwords: Set[str],
    parser_config: ParserConfig,
) -> List[PreparedQuestion]:
    """
    For each candidate card, compute and cache the set of non-stopword tokens from the normalised question text.
    """
    prepared: List[PreparedQuestion] = []
    for card in candidate_cards:
        token_set = tokenise_to_set(card.question_text, stopwords, parser_config)
        prepared.append(
            PreparedQuestion(
                guid=card.guid,
                token_set=token_set,
                non_stopword_count=len(token_set),
                question_token_count=card.question_token_count,
            )
        )
    return prepared


def score_keyword_overlap(
    query_text: str,
    prepared_candidates: List[PreparedQuestion],
    guid_index: Dict[str, Card],
    stopwords: Set[str],
    parser_config: ParserConfig,
    top_k: int = 1,
) -> List[AnswerHit]:
    """
    Score candidates by count of overlapping non-stopword tokens between the query and the candidate question. Apply tie-breakers.
    """
    if top_k < 1:
        top_k = 1

    query_norm = normalise_for_matching(query_text, parser_config)
    query_tokens = tokenise_to_set(query_norm, stopwords, parser_config)

    if not query_tokens or not prepared_candidates:
        return []

    scored: List[Tuple[float, int, int, str]] = []
    for prepared in prepared_candidates:
        overlap_count = len(query_tokens.intersection(prepared.token_set))
        score = float(overlap_count)
        if overlap_count == 0:
            pass
        scored.append((score, overlap_count, prepared.question_token_count, prepared.guid))

    scored.sort(key=lambda t: (-t[0], -t[1], t[2], t[3]))

    hits: List[AnswerHit] = []
    for score, _overlap, question_token_count, guid in scored[:top_k]:
        card = guid_index[guid]
        hits.append(
            AnswerHit(
                guid=guid,
                score=score,
                deck_path=card.deck_path,
                question_preview=short_preview(card.question_text),
            )
        )
    return hits

File: Chatbots/Rust/Source/src/scoring/keyword.rs
use std::collections::HashSet;

use crate::config::ParserConfig;
use crate::data_model::{make_hit, AnswerHit, Card};
use crate::tokenise::{tokenise, tokenise_to_set};

/// A prepared representation of a candidate question for keyword overlap scoring.
#[derive(Debug, Clone)]
pub struct PreparedQuestion {
    pub guid: String,
    pub deck_path: Vec<String>,
    pub question_preview: String,
    pub token_set: HashSet<String>,
    pub token_count: usize,
}

/// Build a prepared index over candidate cards for keyword overlap scoring.
pub fn prepare_keyword_index(
    candidate_cards: &[Card],
    stopword_set: &std::collections::HashSet<String>,
    parser_config: &ParserConfig,
) -> Vec<PreparedQuestion> {
    let mut prepared_list: Vec<PreparedQuestion> = Vec::with_capacity(candidate_cards.len());
    for card in candidate_cards {
        let token_vector = tokenise(&card.question_text, stopword_set, parser_config);
        let token_set = token_vector.iter().cloned().collect::<HashSet<String>>();
        let prepared_question = PreparedQuestion {
            guid: card.guid.clone(),
            deck_path: card.deck_path.clone(),
            question_preview: card.question_text.clone(),
            token_set,
            token_count: token_vector.len(),
        };
        prepared_list.push(prepared_question);
    }
    prepared_list
}

/// Score candidates by keyword overlap (sum of weights = 1 per overlapping token).
/// Tie-breakers:
/// 1) Greater count of overlapping non-stopword tokens
/// 2) Shorter candidate question by token count
/// 3) Lexicographic order of GUID
pub fn score_keyword_overlap(
    query_text: &str,
    prepared_index: &[PreparedQuestion],
    stopword_set: &std::collections::HashSet<String>,
    parser_config: &ParserConfig,
    top_k: usize,
) -> Vec<AnswerHit> {
    let query_token_set = tokenise_to_set(query_text, stopword_set, parser_config);

    let mut scored_hits: Vec<(AnswerHit, usize, usize)> = Vec::with_capacity(prepared_index.len());

    for prepared_question in prepared_index {
        let overlap_count = query_token_set
            .intersection(&prepared_question.token_set)
            .count();

        if overlap_count == 0 {
            // Baseline disallows stopword weights and gives no phrase boosts; zero overlap → score 0, skip.
            continue;
        }

        let score_value = overlap_count as f32; // default weight = 1 per token
        let answer_hit = make_hit(
            prepared_question.guid.clone(),
            prepared_question.deck_path.clone(),
            Some(prepared_question.question_preview.clone()),
            score_value,
        );

        scored_hits.push((
            answer_hit,
            overlap_count,
            prepared_question.token_count,
        ));
    }

    // Sort with tie-breakers: higher score, then higher overlap, then shorter question, then lexicographic GUID
    scored_hits.sort_by(|left, right| {
        right.0.score.partial_cmp(&left.0.score).unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| right.1.cmp(&left.1)) // greater overlap wins
            .then_with(|| left.2.cmp(&right.2)) // shorter question wins
            .then_with(|| left.0.guid.cmp(&right.0.guid))
    });

    scored_hits
        .into_iter()
        .take(top_k)
        .map(|tuple| tuple.0)
        .collect()
}
